src/main/java/org/oscii/detokenize/TrainDetokenizer.java

```diff
@@ -0,0 +1,103 @@
+package org.oscii.detokenize;
+
+import com.google.common.collect.Iterators;
+import edu.stanford.nlp.mt.process.Preprocessor;
+import edu.stanford.nlp.mt.process.de.GermanPreprocessor;
+import edu.stanford.nlp.mt.process.en.EnglishPreprocessor;
+import edu.stanford.nlp.mt.process.es.SpanishPreprocessor;
+import edu.stanford.nlp.mt.process.fr.FrenchPreprocessor;
+import joptsimple.OptionParser;
+import joptsimple.OptionSet;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.zip.GZIPInputStream;
+
+/**
+ * Command-line utility to train a detokenizer.
+ */
+public class TrainDetokenizer {
+    public static void main(String[] args) throws IOException {
+        OptionSet options = parse(args);
+
+        File trainFile = (File) options.valueOf("data");
+        InputStream trainStream = new FileInputStream(trainFile);
+        if (trainFile.getName().endsWith(".gz")) {
+            trainStream = new GZIPInputStream(trainStream);
+        }
+        BufferedReader buffered = new BufferedReader(new InputStreamReader(trainStream, "utf-8"));
+        Iterator<String> trainExamples = buffered.lines().iterator();
+
+        int testSize = (Integer) options.valueOf("testsize");
+        List<String> testExamples = new ArrayList<>(testSize);
+        for (int i = 0; i < testSize; i++) {
+            testExamples.add(trainExamples.next());
+        }
+
+        int trainSize = (Integer) options.valueOf("trainsize");
+        trainExamples = Iterators.limit(trainExamples, trainSize);
+
+        Preprocessor preprocessor = getPreprocessor(options);
+        double regularization = (double) options.valueOf("regularization");
+        Detokenizer detokenizer = Detokenizer.train(preprocessor, regularization, trainExamples);
+
+        System.out.println("Test accuracy: " + detokenizer.evaluate(preprocessor, testExamples.iterator()));
+        if (options.has("errors")) {
+            testExamples.forEach(ex -> {
+                List<String> tokens = Detokenizer.tokenize(preprocessor, ex);
+                String roundTrip = TokenLabel.render(tokens, detokenizer.predictLabels(tokens));
+                if (!ex.equals(roundTrip)) {
+                    System.out.println("Original:  " + ex);
+                    System.out.println("Detoken'd: " + roundTrip);
+                }
+            });
+        }
+
+        File outFile = (File) options.valueOf("out");
+        detokenizer.save(outFile);
+    }
+
+    private static Preprocessor getPreprocessor(OptionSet options) {
+        Preprocessor preprocessor = null;
+        boolean cased = true;
+        switch ((String) options.valueOf("language")) {
+            case "de":
```

Comment:
`tokenize.sh` expects one of `German`, `Spanish`, `English` or `French`; would be great if you could add those as well